README
NAME
     Memory::Process - Perl class to determine actual memory usage.

SYNOPSIS
     use Memory::Process;

     my $m = Memory::Process->new(%params);
     $m->dump;
     $m->record($message, $pid);
     my @report = $m->report;
     my $report = $m->report;
     $m->reset;
     $m->state;

METHODS
  "new"
     my $m = Memory::Process->new(%params);

    Constructor.

    Returns instance of object.

  "dump"
     $m->dump;

    Print report to STDERR.

    Returns return value of print().

  "record"
     $m->record($message, $pid);

    Set record. If message not set, use ''.

    Returns undef.

  "report"
     my @report = $m->report;
     my $report = $m->report;

    Get report.

    In scalar context returns string with report. In array context returns
    array of report lines. First line is title.

  "reset"
     $m->reset;

    Reset records.

    Returns undef.

  "state"
     $m->state;

    Get internal state.

    Each state item consists from:

     - timestamp (in seconds since epoch)
     - message (from record())
     - virtual memory size (in kB)
     - resident set size (in kB)
     - shared memory size (in kB)
     - text size (in kB)
     - data and stack size (in kB)

    Returns reference to array with state items.

EXAMPLE1
     use strict;
     use warnings;

     use Memory::Process;

     # Object.
     my $m = Memory::Process->new;

     # Example process.
     $m->record("Before my big method");
     my $var = ('foo' x 100);
     sleep 1;
     $m->record("After my big method");
     sleep 1;
     $m->record("End");

     # Print report.
     print $m->report."\n";

     # Output like:
     #   time    vsz (  diff)    rss (  diff) shared (  diff)   code (  diff)   data (  diff)
     #      1  19120 (     0)   2464 (     0)   1824 (     0)      8 (     0)   1056 (     0) After my big method
     #      2  19120 (     0)   2464 (     0)   1824 (     0)      8 (     0)   1056 (     0) End

EXAMPLE2
     use strict;
     use warnings;

     use Data::Printer;
     use Memory::Process;

     # Object.
     my $m = Memory::Process->new;

     # Example process.
     $m->record("Before my big method");
     my $var = ('foo' x 100);
     sleep 1;
     $m->record("After my big method");
     sleep 1;
     $m->record("End");

     # Print report.
     my $state_ar = $m->state;

     # Dump out.
     p $state_ar;

     # Output like:
     # \ [
     #     [0] [
     #         [0] 1445941214,
     #         [1] "Before my big method",
     #         [2] 33712,
     #         [3] 7956,
     #         [4] 3876,
     #         [5] 8,
     #         [6] 4564
     #     ],
     #     [1] [
     #         [0] 1445941215,
     #         [1] "After my big method",
     #         [2] 33712,
     #         [3] 7956,
     #         [4] 3876,
     #         [5] 8,
     #         [6] 4564
     #     ],
     #     [2] [
     #         [0] 1445941216,
     #         [1] "End",
     #         [2] 33712,
     #         [3] 7956,
     #         [4] 3876,
     #         [5] 8,
     #         [6] 4564
     #     ]
     # ]

DEPENDENCIES
    Memory::Usage, Readonly.

SEE ALSO
    Memory::Stats
        Memory Usage Consumption of your process

    Memory::Usage
        Tools to determine actual memory usage

REPOSITORY
    <https://github.com/michal-josef-spacek/Memory-Process>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz/>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07

0.06 2023-04-01T13:13:24+02:00
 - Fix supported OSes.

0.05 2023-03-30T15:44:51+02:00
 - Add names for example files.
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Fix repository url.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Rewrite doc to use =head2.
 - Rewrite tests from IO::CaptureOutput to Capture::Tiny.
 - Sort methods in SYNOPSIS section in doc.
 - Update author name.
 - Update copyright years.
 - Update Module::Install to 1.19 version.

0.04 2015-10-29T20:04:23+01:00
 - Add reset() method.
 - Fix requires.
 - Improve tests.

0.03 2015-10-27T12:12:48+01:00
 - Add EXAMPLE2 example to doc.
 - Add documentation for state() method.
 - Add tests.
 - Fix report() method in doc.
 - Implement state() method, which is better than Memory::Usage->state().
 - Implement record() method, which can has undefined message.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SEE ALSO section in doc.
 - Improve report() method return values.
 - Move author tests to xt/ directory.
 - Update Module::Install to 1.16 version.
 - Update copyright years in LICENSE file.

0.02 2014-11-08T08:29:33+01:00
 - Fix dump() across Memory::Usage->dump().
 - Fix report() across Memory::Usage->report() in scalar mode.

0.01 2014-11-07T12:26:38+01:00
 - First version.


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Memory process reporting.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.05;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Memory-Process';
readme_from 'Process.pm';
recursive_author_tests('xt');
requires 'Memory::Usage' => 0;
requires 'Readonly' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://rt.cpan.org/Public/Dist/Display.html'.
	'?Name=Memory-Process';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Memory-Process';
resources 'repository' => 'git://github.com/michal-josef-spacek/Memory-Process';
test_requires 'Capture::Tiny' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.07';

if ($^O eq 'freebsd' || $^O eq 'MSWin32' || $^O eq 'openbsd' || $^O eq 'solaris'
	|| $^O eq 'darwin') {

	die "System '$^O' doesn't supported.";
}

# Run.
WriteAll();


Process.pm
package Memory::Process;

use base qw(Memory::Usage);
use strict;
use warnings;

use Readonly;

# Constants.
Readonly::Scalar our $EMPTY_STR => q{};

our $VERSION = 0.07;

# Record.
sub record {
	my ($self, $message, $pid) = @_;
	if (! defined $message) {
		$message = $EMPTY_STR;
	}
	return $self->SUPER::record($message, $pid);
}

# Print report to STDERR.
sub dump {
	my $self = shift;
	return print STDERR scalar $self->report;
}

# Get report.
sub report {
	my $self = shift;
	my $report = $self->SUPER::report;
	my @report_full = split m/\n/ms, $report;
	my @report = ();
	if (@report_full > 2) {
		@report = ($report_full[0], $report_full[-2], $report_full[-1]);
	};
	my $report_scalar = (join "\n", @report);
	if ($report_scalar ne $EMPTY_STR) {
		$report_scalar .= "\n";
	}
	return wantarray ? @report : $report_scalar;
}

# Reset records.
sub reset {
	my $self = shift;
	@{$self} = ();
	return;
}

# Get state.
sub state {
	my $self = shift;
	return [@{$self}];
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

 Memory::Process - Perl class to determine actual memory usage.

=head1 SYNOPSIS

 use Memory::Process;

 my $m = Memory::Process->new(%params);
 $m->dump;
 $m->record($message, $pid);
 my @report = $m->report;
 my $report = $m->report;
 $m->reset;
 $m->state;

=head1 METHODS

=head2 C<new>

 my $m = Memory::Process->new(%params);

Constructor.

Returns instance of object.

=head2 C<dump>

 $m->dump;

Print report to STDERR.

Returns return value of print().

=head2 C<record>

 $m->record($message, $pid);

Set record.
If message not set, use ''.

Returns undef.

=head2 C<report>

 my @report = $m->report;
 my $report = $m->report;

Get report.

In scalar context returns string with report.
In array context returns array of report lines.
First line is title.

=head2 C<reset>

 $m->reset;

Reset records.

Returns undef.

=head2 C<state>

 $m->state;

Get internal state.

Each state item consists from:

 - timestamp (in seconds since epoch)
 - message (from record())
 - virtual memory size (in kB)
 - resident set size (in kB)
 - shared memory size (in kB)
 - text size (in kB)
 - data and stack size (in kB)

Returns reference to array with state items.

=head1 EXAMPLE1

=for comment filename=run_process_and_get_report.pl

 use strict;
 use warnings;

 use Memory::Process;

 # Object.
 my $m = Memory::Process->new;

 # Example process.
 $m->record("Before my big method");
 my $var = ('foo' x 100);
 sleep 1;
 $m->record("After my big method");
 sleep 1;
 $m->record("End");

 # Print report.
 print $m->report."\n";

 # Output like:
 #   time    vsz (  diff)    rss (  diff) shared (  diff)   code (  diff)   data (  diff)
 #      1  19120 (     0)   2464 (     0)   1824 (     0)      8 (     0)   1056 (     0) After my big method
 #      2  19120 (     0)   2464 (     0)   1824 (     0)      8 (     0)   1056 (     0) End

=head1 EXAMPLE2

=for comment filename=run_process_and_get_state.pl

 use strict;
 use warnings;

 use Data::Printer;
 use Memory::Process;

 # Object.
 my $m = Memory::Process->new;

 # Example process.
 $m->record("Before my big method");
 my $var = ('foo' x 100);
 sleep 1;
 $m->record("After my big method");
 sleep 1;
 $m->record("End");

 # Print report.
 my $state_ar = $m->state;

 # Dump out.
 p $state_ar;

 # Output like:
 # \ [
 #     [0] [
 #         [0] 1445941214,
 #         [1] "Before my big method",
 #         [2] 33712,
 #         [3] 7956,
 #         [4] 3876,
 #         [5] 8,
 #         [6] 4564
 #     ],
 #     [1] [
 #         [0] 1445941215,
 #         [1] "After my big method",
 #         [2] 33712,
 #         [3] 7956,
 #         [4] 3876,
 #         [5] 8,
 #         [6] 4564
 #     ],
 #     [2] [
 #         [0] 1445941216,
 #         [1] "End",
 #         [2] 33712,
 #         [3] 7956,
 #         [4] 3876,
 #         [5] 8,
 #         [6] 4564
 #     ]
 # ]

=head1 DEPENDENCIES

L<Memory::Usage>,
L<Readonly>.

=head1 SEE ALSO

=over

=item L<Memory::Stats>

Memory Usage Consumption of your process

=item L<Memory::Usage>

Tools to determine actual memory usage

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Memory-Process>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz/>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut
